src/client/pmix_client.c

```diff
@@ -570,56 +561,81 @@ pmix_status_t PMIx_Put(pmix_scope_t scope, const char key[], pmix_value_t *val)
 
     /* pack the cache that matches the scope - global scope needs
      * to go into both local and remote caches */
-    if (PMIX_LOCAL == scope || PMIX_GLOBAL == scope) {
+    if (PMIX_LOCAL == cb->scope || PMIX_GLOBAL == cb->scope) {
         if (NULL == pmix_globals.cache_local) {
             pmix_globals.cache_local = PMIX_NEW(pmix_buffer_t);
         }
         pmix_output_verbose(2, pmix_globals.debug_output,
                             "pmix: put %s data for key %s in local cache",
-                            key, (PMIX_GLOBAL == scope) ? "global" : "local");
+                            cb->key, (PMIX_GLOBAL == cb->scope) ? "global" : "local");
         if (PMIX_SUCCESS != (rc = pmix_bfrop.pack(pmix_globals.cache_local, kv, 1, PMIX_KVAL))) {
             PMIX_ERROR_LOG(rc);
         }
     }
 
-    if (PMIX_REMOTE == scope || PMIX_GLOBAL == scope) {
+    if (PMIX_REMOTE == cb->scope || PMIX_GLOBAL == cb->scope) {
         if (NULL == pmix_globals.cache_remote) {
             pmix_globals.cache_remote = PMIX_NEW(pmix_buffer_t);
         }
         pmix_output_verbose(2, pmix_globals.debug_output,
                             "pmix: put %s data for key %s in remote cache",
-                            key, (PMIX_GLOBAL == scope) ? "global" : "remote");
+                            cb->key, (PMIX_GLOBAL == cb->scope) ? "global" : "remote");
         if (PMIX_SUCCESS != (rc = pmix_bfrop.pack(pmix_globals.cache_remote, kv, 1, PMIX_KVAL))) {
             PMIX_ERROR_LOG(rc);
         }
     }
 
+  done:
     PMIX_RELEASE(kv);  // maintain accounting
+    cb->pstatus = rc;
+    cb->active = false;
+}
+
+pmix_status_t PMIx_Put(pmix_scope_t scope, const char key[], pmix_value_t *val)
+{
+    pmix_cb_t *cb;
+    pmix_status_t rc;
+
+    pmix_output_verbose(2, pmix_globals.debug_output,
+                        "pmix: executing put for key %s type %d",
+                        key, val->type);
+
+    if (pmix_globals.init_cntr <= 0) {
+        return PMIX_ERR_INIT;
+    }
+
+    /* create a callback object */
+    cb = PMIX_NEW(pmix_cb_t);
+    cb->active = true;
+    cb->scope = scope;
+    cb->key = (char*)key;
```

Comment:
@rhc54 the current version segfaults. So we need to solve it somehow. 
My suggestion is:
1. strdup key in PMIx_Put.
2. in _putfn pass it further to the kv->key data structure:
```
kv->key = cb->key;
cb->key = NULL;  // protect the data
```
you will need to `NULL` `cb->key` because `pmix_kval_t` destructor also free()'s the key if it is not zero. So we **must** have key malloc'ed.

this was my original suggestion. Sorry if it wasn't clear.